Introduce new semantic conventions for CloudEvents (#1978)
open-telemetry/opentelemetry-specification: specification/trace/semantic_conventions/cloudevents.md

to provide interoperability across services, platforms and systems.

For more information and background on the history of how CloudEvents came to
fruition, consult the
[CloudEvents Primer](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/primer.md)
document.

CloudEvents are typically used in distributed, event-based systems.
Tracing such event-based systems can be a challenging task to achieve.
Many components can produce and consume events and such events can trigger
the dispatch of further events, all in an asynchronous manner.

The semantic conventions below, define how to model the different stages of a
CloudEvent in event-based systems. It covers creation, processing,
context propagation between producer and consumer and how to enrich spans
with CloudEvent-specific attributes.

With the proposed model, it is possible to have an overview of everything
that happened as the result of an event. One can, for example, answer the
following questions:

- What components in a system reacted to an event
- What further events were sent due to an incoming event
- Which event caused the exception

## Overview

> To be individually traceable, each CloudEvent has to have its own trace context.

Reviewer: What does "individually traceable" mean?

Author: The answer above goes into more details, but an event can go over multiple hops in a distributed system and one would want to trace every action/reaction that happened due to an event being sent. In order to do that, each event must have its own context, that "travels" together with it and other traces can link to it.

The trace context is populated in the event using the
[CloudEvents Distributed Tracing Extension](https://github.com/cloudevents/spec/blob/v1.0.1/extensions/distributed-tracing.md).

Once the trace context is set on the event, it MUST not be modified.

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Spans](#spans)
  * [Creation](#creation)
  * [Processing](#processing)
- [Attributes](#attributes)

<!-- tocstop -->

## Spans

### Creation

> Instrumentation SHOULD create a new span and populate the

Reviewer: Is a new span created for each cloudevent? What's the duration of the span?

Author: Yes, a new span should be created for each new event. There will be essentially no duration as this span is more a way of creating a "snapshot" that this happen so this context can travel along with the event and can be linked later by any involved component.

> [CloudEvents Distributed Tracing Extension](https://github.com/cloudevents/spec/blob/v1.0.1/extensions/distributed-tracing.md)

Reviewer: Do I understand correctly that the purpose of this is to capture the context of the execution so that when the cloudevent is propagated elsewhere this context is preserved and can be continued on? This is essentially the context propagation mechanism and the cloudevent is the medium through which the context propagates. Is this true?

Author: Yes, you understood it correctly.

A CloudEvent can be sent directly from producer to consumer. In this manner the customary parent-child trace model works well and we don't need extra context propagation. But, CloudEvents also often go through many hops until it reaches a consumer, specially in distributed systems. This makes it hard to create a trace that "makes sense". This challenge is similar (if not the same) as the ones we face with messaging systems. 

The CloudEvents spec already offers as mentioned above an "extension" where a span context can be added to the event as a way of preserving the initial create context. This specification then takes uses of it and offers guidelines on how/when to populate it and how to consume it in order to continue the trace in a consistent way. 

To help exemplify, consider a scenario like this:

![image](https://user-images.githubusercontent.com/5938087/154439136-6ae4e1f2-682f-4fb8-9969-180a82f41761.png)

Today, we will get a trace like this (given there's some auto-instrumentation configured, e.g. http):
![image](https://user-images.githubusercontent.com/5938087/154440965-2a91370c-ed9d-45d1-92e7-f5462b690112.png)

Essentially you have 2 traces that you can't correlate to each other. Now, here's a trace that implements the proposed conventions here:
![image](https://user-images.githubusercontent.com/5938087/154440598-f502318b-ec97-4c71-b266-d3e260b44ec1.png)

The two "red" spans are the `Create` and `Process` this conventions introduces. This gives a chance of connecting producer and consumer(s), basically enabling users to have a "map" of everything that happened after an event was sent.
This model also creates a trace that is independent of any protocols.

To also help clarify, this way of using the event as a medium to propagate the initial create context is also similar to what we are discussing for messaging. In messaging we will also attach the initial create context to the message in some way (message metadata, etc).

Reviewer: This explanation and the diagrams are very useful. I think it is worth adding as a section to the conventions document. It makes the intent clear.

Author: I'll try to move some of it to the document. I'm not sure how to handle images in here though.. I guess I can try reproducing it with some fancy ASCII art.

Reviewer: Either ASCII art or add pngs. We do that elsewhere in specs already, e.g.: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/library-guidelines.md

Author: @tigrannajaryan I added the diagrams + modified the sections a bit to include the explanation from this thread.

# CloudEvents

**Status**: [Experimental](../../document-status.md)

This specification defines attributes for [CloudEvents](https://cloudevents.io/).

> The attributes described in this document are not specific to a particular operation but rather generic. They may be used in any Span they apply to.

Reviewer 2: What about span events? Rather not, right?

Author: Yeah I'd say not at this point. Maybe it will make sense later on though.

# CloudEvents

**Status**: [Experimental](../../document-status.md)

This specification defines attributes for [CloudEvents](https://cloudevents.io/).
The attributes described in this document are not specific to a particular operation but rather generic. They may be used in any Span they apply to.

At this time, this specification does not cover:

> - How and when Spans are created for CloudEvents

Reviewer 3: and still, I don't understand the benefit of defining a spec that's not possible to implement. Who does it help?

Author: As per discussion above, I removed it now. Now it simply has the attributes. I understand and agree with your points, but it seems we are a bit far to reaching an conclusion. It seems we will only be "unblocked" once we figure out the propagation portion. I think what @pyohannes said of capturing these missing things in an issue makes more sense. What do you think?

Reviewer 4: I'll reserve a slot in this Thursday's messaging workgroup so we can align on this (whether to add instrumentation guidance besides just attributes).

I had a quick chat with @lmolkova, she has a strong interest in having this instrumentation guidance worked out, and she also offered to contribute to this work.

- Transport aspects of sending and receiving CloudEvents

The above points are being worked on in the messaging semantic conventions.

> This document will be updated with further guidance once the messaging semantic conventions reaches a stable state.

Reviewer 2: Do we intend to wait until after the messaging semantic conventions reach "stable"? No semantic convention has ever reached that state yet (except service.name).

Author: Yeah probably not, but I couldn't find a better way to explain it. =/

I wanted to express that some concepts are "overlapping"  and that more details with be added when we figure out things in the messaging semantic conventions. (e.g.** how to propagate context? Should users use CloudEvents extensions, or should it be taken care of by the transport instrumentation, see Liudmila's concerns) 

I had this version in a draft, but I'm not sure it's better:
```
This document will be updated with further guidance once the messaging working group defines a clear guidance on how CloudEvents related with messaging systems.
```

Reviewer 4: I'm fine leaving this statement out. So let's not elaborate on what's not currently covered and when it will (prospectively) be covered. Again, maybe this is better captured in an issue.

Author: Alright - I removed both statements. Does it also makes sense then to remove this portion:

> At this time, this specification does not cover:

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Attributes](#attributes)

<!-- tocstop -->

## Attributes

<!-- semconv cloudevents -->
| Attribute  | Type | Description  | Examples  | Required |
|---|---|---|---|---|

> | `cloudevents.event_id` | string | The [event_id](https://github.com/cloudevents/spec/blob/master/spec.md#id) identifies the event within the scope of a producer. | `123e4567-e89b-12d3-a456-426614174000`; `producer-1` | No |

Reviewer 4: It's not clear to me what "identifies the event within the scope of a producer" means. Aren't the id and the source supposed to help uniquely identify an event also on consumer side?

Author: From the CloudEvents spec, the `id` field lists some constraints, one of them being:

> MUST be unique within the scope of the producer

But you are right, it is a bit confusing. I've changed to only say it uniquely identifies the event, and added a `note` about the `source` + `id` combination as well.

| `cloudevents.source` | string | The [source](https://github.com/cloudevents/spec/blob/master/spec.md#source-1) identifies the context in which an event happened. | `https://github.com/cloudevents`; `/cloudevents/spec/pull/123`; `my-service` | No |
| `cloudevents.spec_version` | string | The [version of the CloudEvents specification](https://github.com/cloudevents/spec/blob/master/spec.md#specversion) which the event uses. | `1.0` | No |
| `cloudevents.event_type` | string | The [event_type](https://github.com/cloudevents/spec/blob/master/spec.md#type) contains a value describing the type of event related to the originating occurrence. | `com.github.pull_request.opened`; `com.example.object.deleted.v2` | No |

> | `cloudevents.subject` | string | The [subject](https://github.com/cloudevents/spec/blob/master/spec.md#subject) of the event in the context of the event producer (identified by source). | `mynewfile.jpg` | No |

Reviewer 4: Just a question about consistent naming: as far as I understand, all those attributes pertain to a single event. Only two of them (`event_id` and `event_type`) have an `event_` prefix, others (`source` or `subject`) don't.

Do you think it makes sense to make this consistent?

Author: I think it makes sense to add the `event_` prefix to all of them. I initially copied them from the CloudEvents Go SDK and only added the prefix to `id` and `type` because they were "too general" and might be confused with other conventions. Changed it now.

# CloudEvents

> This document defines how to describe sending and receiving [CloudEvents](https://cloudevents.io/) with spans.

Reviewer 4: I wonder if we could restrict these conventions to just defining attributes, and leaving it open how they are added to spans?

I'm not sure if having separate spans for CloudEvents makes sense in all cases.

Reviewer 2: I also think that in some cases you'd want to add the span attributes to the span of the operation that is described by the event (e.g. for `com.github.pull_request.opened` you would add the attributes to the span that traces opening the PR).

Author: Yeah I agree and it was what I tried to explain in the opening statement, but it's not good. Would you have an idea how we can communicate this more clearly?

Reviewer 3: Since we are separating creation and publishing in the new messaging SDKs, CloudEvent span in this case is creation, and HTTP POST  is publishing. Cloud events spans seem to be essential at least for this SDK and attributes should not be stamped on any other spans - there is no technical solution that would even allow that except hacky manual instrumentation colliding with an automatic one. 

They would be duplicates with messaging SDKs that use cloud events (have 'create' span too and propagate context differently) - but that's what we have to solve in messaging spec.
I think if users turn cloud events instrumentation on (or don't turn it off), they should have cloud events spans regardless of messaging layer.

Reviewer 2: > CloudEvent span in this case is creation

I disagree. I was only there in the first SIG meeting, but there "creation" was defined as the logical send operation, so it would include retries, etc. but not the creation of the data from which the message is formed.

Reviewer 2: With such a broad definition of "creation", all business logic that happens to also notify something via message about the result would be traced with a "message creation" span.

Reviewer 3: > "creation" was defined as the logical send operation

I don't think so, Creation is a separate step according to the [draft](https://github.com/open-telemetry/oteps/blob/922295985f2f1c0e440d4fe7c91f34f43e5e87e3/text/trace/0157-messaging-semantic-conventions.md)

Since you can send multiple cloud events over whatever protocol in a batch and CloudEvent has an individual context, you have to create a span for it.

Reviewer 3: and here's the thread that clarifies it https://github.com/open-telemetry/oteps/pull/173#discussion_r701762856

Reviewer 3: > sending and receiving

I suggest changing it to creation and consumption (sending and receiving are independent at least for other SDKs)

Author: I read the thread you linked, I think it makes sense. But there it's said:

> "Create" (as well as "Process") are agnostic of the underlying protocol (AMQP, MQTT, HTTP), whereas all the other stages are not.

So probably we should use here also create and process to be consistent.

Reviewer 3: sure, I agree to use terminology that's used in the OTEP and 'process' seems to be more precise.

Reviewer 4: > Since we are separating creation and publishing in the new messaging spec, CloudEvent span in this case is creation, and HTTP POST is publishing.

We didn't agree on any span structure yet, but we decided to treat creation and publishing as different conceptual stages.

I agree with @lmolkova we should have a clear guidance of when and whether to create dedicated CloudEvents span, but I think currently we don't have currently alignment on that.

I propose we just specify the attributes for now, maybe with a similar disclaimer than [for general span attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/span-general.md#general-attributes):

> The attributes described in this section are not specific to a particular operation but rather generic. They may be used in any Span they apply to.

We might want to add a hint that this will be specified in more detail in the future  (when we work out the conventions for messaging).

Reviewer 3: > I propose we just specify the attributes for now, maybe with a similar disclaimer than for general span attributes:

With spec that intentionally avoids specifying which spans should be created, I can't take it as a library owner and implement it. Backends can't take it and build experiences on it. General attributes beauty that they can go an almost any span, these can only go on something related to cloud events.

This spec (among others) is experimental and if messaging spec would contradict it, we can come back and change it, but leaving it intentionally vague does not seem to help.

Reviewer 2: >> @Oberon00:  I was only there in the first SIG meeting, but there "creation" was defined as the logical send operation, so it would include retries, etc. but not the creation of the data from which the message is formed.
>
> @lmolkova: I don't think so, Creation is a separate step according to the draft
>
> and here's the thread that clarifies it https://github.com/open-telemetry/oteps/pull/173#discussion_r701762856

But in this thread I said exactly that at https://github.com/open-telemetry/oteps/pull/173#discussion_r702089063:

>> @Oberon00: From the WG meeting it sounded like Create is actually supposed to mean "logical publish operation" (potentially with retries) while Publish is the "physical publish operation".
>
> @pyohannes: That's my understanding, seems we have to make this clearer in the document.

Author: Hey all. I read the thread in the OTEP and although it seems we reached an agreement that it's helpful to have the create/publish stages separate, we still don't have a clear way on how we want to do it, as we can see here in this thread.

With that in mind, I took a few suggestions from @lmolkova and @pyohannes and I adapted the document to explicit say this are "generic" attributes and that further guidance will be added once we figure out the details on how CloudEvents will relate with messaging. Let me know your thoughts on this update.

> The transport involved in sending/receiving CloudEvents is represented by a separate span, depending on the underlying protocol used.

Reviewer 4: What does "depending on the underlying protocol used" mean? Does it mean some underlying protocols require a separate span to be created, others do not?

Author: I wanted to basically say what you mentioned below: 

> I wonder if we could restrict these conventions to just defining attributes, and leaving it open how they are added to spans?

What I wanted to say is that you can use the attributes here to create spans if you want but all the other details of propagation and probably separate spans that need to be created when actually sending/receiving the event is a different topic. I'm having a hard time to produce words to explain this 😅.

An actual example:

```go
event := cloudevents.NewEvent()

// create a span for this operation, and another one maybe be created, depending if the transport is instrumented
ctx, span := o.tracer.Start(ctx, 
		"CloudEvents " + e.Context.GetType(),
		trace.WithSpanKind(trace.SpanKindProducer),
		trace.WithAttributes(...)) // with the attributes from OTel semconv
		
client, _ := client.New(t, copt...)
// Another span might be created here, depending if the transport being used is instrumented
client.Send(ctx, event) 
span.End()
```

Reviewer 3: > The transport involved in sending/receiving CloudEvents

I suggest this spec does not cover this subject at all and it's up to messaging spec to define

Author: Sounds good, but I'm not sure how express this here in this spec. Do you have a suggestion?

Reviewer 3: I'd suggest something along the lines of "This specification covers creation and consumption of cloud events only and does not cover transport aspects of publishing and receiving cloud events.  Please refer to the messaging semantic conventions that will cover creation, consumption, and transport. 
The integration between CloudEvents and messaging libraries tracing instrumentation is not defined yet. Instrumentations should follow either CloudEvents convention or Messaging convention, but not both"

Author: Sorry I read after your comment below about using "create", but I think consumption should be process? As it's in the thread in the OTEP you linked. Thanks for the suggestion I'll try to work something up!

# CloudEvents

**Status**: [Experimental](../../document-status.md)

This specification defines semantic conventions for [CloudEvents](https://cloudevents.io/). It covers creation, processing and context propagation between producer and consumer. It does not cover transport aspects of publishing and receiving cloud events.

## Overview

To be individually traceable, each CloudEvent has to have its own trace context. The trace context is populated in the event using the [Distributed Tracing Extension](https://github.com/cloudevents/spec/blob/v1.0.1/extensions/distributed-tracing.md).

Once the trace context is set on the event, it MUST not be modified.

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Spans](#spans)
  * [Creation](#creation)
  * [Processing](#processing)
- [Attributes](#attributes)

<!-- tocstop -->

## Spans

### Creation

> Instrumentation SHOULD create a new span and populate the [Distributed Tracing Extension](https://github.com/cloudevents/spec/blob/v1.0.1/extensions/distributed-tracing.md) on the event. This applies when:

Reviewer 4: From discussions in the messaging workgroup: we should add a link to this document in the CloudEvents spec, to make clear where the DT extension is used.

Author: I have a PR in the pipeline. I'll open it later today or, at the latest tomorrow (against the CloudEvents spec).

Author: PR is there: https://github.com/cloudevents/spec/pull/912

Reviewer 5: ```suggestion
Instrumentation SHOULD create a new span and populate the [CloudEvents Distributed Tracing Extension](https://github.com/cloudevents/spec/blob/v1.0.1/extensions/distributed-tracing.md) on the event. This applies when:
```

Author: Done.

- A CloudEvent is created by the instrumented library. It may be impossible or impractical to create the Span during event creation (e.g. inside the constructor or in a factory method) - instrumentation MAY create the Span later, when passing the event to the transport layer.
- A CloudEvent is created outside of the instrumented library (e.g. directly constructed by the application owner, without calling a constructor or factory method), and passed without the Distributed Tracing Extension populated.

In case a CloudEvent is passed to the instrumented library with the Distributed Tracing Extension populated, instrumentation MUST NOT create a span and MUST NOT modify the Distributed Tracing Extension on the event.

**Span name:** `CloudEvents Create <event_type>`
**Span kind:** PRODUCER

### Processing

When instrumented library supports processing of a single CloudEvent, instrumentation SHOULD create a new span to trace it. Instrumentation SHOULD set the remote trace context from the Distributed Tracing Extension as a link on the processing span.

**Span name:** `CloudEvents Process <event_type>`
**Span kind:** CONSUMER

> Note: CloudEvents processing does not follow a common pattern. Some "CloudEvents-enabled" libraries expose handlers for processing (e.g.  [CloudEvents Go SDK](https://github.com/cloudevents/sdk-go#receive-your-first-cloudevent)) and can auto-instrument the processing calls. Alternatively, a library might offer a handler, but the user chooses not to use it, or a library does not offer one at all. In such cases, the instrumentation responsibility lies with the application owner.

Reviewer 4: From discussions in the messaging workgroup: this section could be more authoritative and tell SDKs what to do. E. g.: SDKs SHOULD preferably auto-instrument or alternatively provide a handler.

Author: I changed the paragraph a bit, we can go over it during this week's messaging wg meeting.

# CloudEvents

**Status**: [Experimental](../../document-status.md)

This specification defines semantic conventions for [CloudEvents](https://cloudevents.io/). It covers creation, processing and context propagation between producer and consumer. It does not cover transport aspects of publishing and receiving cloud events.

## Overview

> To be individually traceable, each CloudEvent has to have its own trace context. The trace context is populated in the event using the [Distributed Tracing Extension](https://github.com/cloudevents/spec/blob/v1.0.1/extensions/distributed-tracing.md).

Reviewer 5: ```suggestion
To be individually traceable, each CloudEvent has to have its own trace context.
The trace context is populated in the event using the [CloudEvents Distributed Tracing Extension](https://github.com/cloudevents/spec/blob/v1.0.1/extensions/distributed-tracing.md).
```
To make clear this is provided by CloudEvents, not OTel.

Author: Makes sense. I also broke the long lines to make it easier to review.

Once the trace context is set on the event, it MUST not be modified.

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Spans](#spans)
  * [Creation](#creation)
  * [Processing](#processing)
- [Instrumentation](#instrumentation)
- [Attributes](#attributes)

<!-- tocstop -->

## Spans

### Creation

Instrumentation SHOULD create a new span and populate the [Distributed Tracing Extension](https://github.com/cloudevents/spec/blob/v1.0.1/extensions/distributed-tracing.md) on the event. This applies when:

> - A CloudEvent is created by the instrumented library. It may be impossible or impractical to create the Span during event creation (e.g. inside the constructor or in a factory method) - instrumentation MAY create the Span later, when passing the event to the transport layer.

Reviewer 5: ```suggestion
- A CloudEvent is created by the instrumented library. It may be impossible or impractical to create the Span during event creation (e.g. inside the constructor or in a factory method), so instrumentation MAY create the Span later, when passing the event to the transport layer.
```

- A CloudEvent is created outside of the instrumented library (e.g. directly constructed by the application owner, without calling a constructor or factory method), and passed without the Distributed Tracing Extension populated.

> In case a CloudEvent is passed to the instrumented library with the Distributed Tracing Extension populated, instrumentation MUST NOT create a span and MUST NOT modify the Distributed Tracing Extension on the event.

Reviewer 5: ```suggestion
In case a CloudEvent is passed to the instrumented library with the Distributed Tracing Extension already populated, instrumentation MUST NOT create a span and MUST NOT modify the Distributed Tracing Extension on the event.
```

**Span name:** `CloudEvents Create <event_type>`
**Span kind:** PRODUCER

### Processing

> When instrumented library supports processing of a single CloudEvent, instrumentation SHOULD create a new span to trace it. Instrumentation SHOULD set the remote trace context from the Distributed Tracing Extension as a link on the processing span.

Reviewer 5: ```suggestion
When an instrumented library supports processing of a single CloudEvent, instrumentation SHOULD create a new span to trace it.
Instrumentation SHOULD set the remote trace context from the Distributed Tracing Extension as a link on the processing span.
```

**Span name:** `CloudEvents Process <event_type>`
**Span kind:** CONSUMER

## Instrumentation

CloudEvents libraries SHOULD have built-in mechanisms to make the aforementioned
Span operations (creation, extraction, and injection of context into events)
transparent for application owners. When auto-instrumentation
is not feasible, CloudEvents libraries SHOULD offer such mechanisms in a form that

> application owners can "plug in" to apply the Span operations above by themselves.

Reviewer 5: The part "CloudEvents libraries SHOULD" should rather go into https://github.com/cloudevents/spec than the OTel spec I think.

Author: I tend to agree as well. I think this should be a section in the  https://github.com/cloudevents/spec/blob/v1.0.1/extensions/distributed-tracing.md. Looping in @pyohannes, @lmolkova. What do you think? I vote to remote this from here as well. I spoke with the maitainer of the JS CloudEvents SDK and he also feel this is not the place to be. As a SDK authors people will look at the CloudEvents spec on what the SDK should do, and not here.

## Attributes

The following attributes are applicable to creation and processing Spans.

<!-- semconv cloudevents -->
| Attribute  | Type | Description  | Examples  | Required |
|---|---|---|---|---|
| `cloudevents.event_id` | string | The [event_id](https://github.com/cloudevents/spec/blob/master/spec.md#id) uniquely identifies the event. [1] | `123e4567-e89b-12d3-a456-426614174000`; `0001` | No |
| `cloudevents.event_source` | string | The [source](https://github.com/cloudevents/spec/blob/master/spec.md#source-1) identifies the context in which an event happened. | `https://github.com/cloudevents`; `/cloudevents/spec/pull/123`; `my-service` | No |
| `cloudevents.event_spec_version` | string | The [version of the CloudEvents specification](https://github.com/cloudevents/spec/blob/master/spec.md#specversion) which the event uses. | `1.0` | No |
| `cloudevents.event_type` | string | The [event_type](https://github.com/cloudevents/spec/blob/master/spec.md#type) contains a value describing the type of event related to the originating occurrence. | `com.github.pull_request.opened`; `com.example.object.deleted.v2` | No |

> | `cloudevents.event_subject` | string | The [subject](https://github.com/cloudevents/spec/blob/master/spec.md#subject) of the event in the context of the event producer (identified by source). | `mynewfile.jpg` | No |

Reviewer 5: https://github.com/cloudevents/spec renamed master to main and restructured directories so the links here are broken.

Author: Thanks! I fixed it and used the `v1.0.1` default branch now.

> # CloudEvents

Reviewer: Is it possible to add some context/background to this doc? A short paragraph describing what is a cloudevent and why and when (in what scenarios) one would want to record it in span attributes?
I don't know anything about CloudEvents and tried reading https://cloudevents.io/ and still can't figure out why it needs to be recorded on a span.

Author: Hi @tigrannajaryan I pushed some changes where I tried to answer your questions. Let me know if this helps clarifies things.

Author: I should also mention that this is still experimental and many things we came up in this PR (we, as the Messaging working group) are being used in stabilizing semantic conventions for messaging systems (https://github.com/open-telemetry/oteps/pull/192). 

As we progress with messaging systems semantic conventions, this document will change to reflect that. Specially in things like context propagation.

Author: Resolving this as most things got answered in this thread https://github.com/open-telemetry/opentelemetry-specification/pull/1978#discussion_r808532035

creation (e.g. inside the constructor or in a factory method),
so instrumentation MAY create the Span later, when passing the event to the transport layer.
- A CloudEvent is created outside of the instrumented library
(e.g. directly constructed by the application owner, without calling a constructor or factory method),
and passed without the Distributed Tracing Extension populated.

In case a CloudEvent is passed to the instrumented library with the
Distributed Tracing Extension already populated, instrumentation MUST NOT create
a span and MUST NOT modify the Distributed Tracing Extension on the event.

**Span name:** `CloudEvents Create <event_type>`

**Span kind:** PRODUCER

### Processing

When an instrumented library supports processing of a single CloudEvent,
instrumentation SHOULD create a new span to trace it.

Instrumentation SHOULD set the remote trace context from the
Distributed Tracing Extension as a link on the processing span.

**Span name:** `CloudEvents Process <event_type>`

**Span kind:** CONSUMER

## Attributes

The following attributes are applicable to creation and processing Spans.

<!-- semconv cloudevents -->

> | Attribute  | Type | Description  | Examples  | Required |

Reviewer: Are all of the attributes optional? Typically we have at least one required attribute in a set of attributes that describe a concept. The presence of such attribute makes it possible to detect that the span is about that particular concept.
If there is an attribute that you know will always be available for a cloudevent it would be useful to mark that attribute as Required.

Author: Thanks for pointing this out. There are several attributes in the CloudEvent specification that are required in the event. (https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#required-attributes). I added these here as required as well.

# CloudEvents

**Status**: [Experimental](../../document-status.md)

This specification defines attributes for [CloudEvents](https://cloudevents.io/).
The attributes described in this document are not specific to a particular operation but rather generic. They may be used in any Span they apply to.

At this time, this specification does not cover:

- How and when Spans are created for CloudEvents

- Transport aspects of sending and receiving CloudEvents

> The above points are being worked on in the messaging semantic conventions.

Reviewer 2: I would not even mention messaging. I'm still not sure if there is much overlap.

Author: From the discussion above, I got that most of the details will come from the messaging WG/conventions, no? For example, if a Span should be created when creating CloudEvents. So I thought of mentioned it as @pyohannes suggested:

> We might want to add a hint that this will be specified in more detail in the future (when we work out the conventions for messaging).

Reviewer 2: > From the discussion above, I got that most of the details will come from the messaging WG/conventions, no? For example, if a Span should be created when creating CloudEvents.

Why do we see CloudEvents as a messaging system? Does messaging SIG also consider general event notification systems in scope? I guess it makes sense then, fine with me then.

Reviewer 3: messaging systems carry a payload that historically is a blob or arbitrary data, also they all have different metadata. CloudEvents are intended to replace the arbitrary blob and some metadata, they are designed to be used with messaging systems https://github.com/cloudevents/spec#cloudevents-documents

The intersection between this spec and messaging spec is the trace context. They'd both define when the context is created, how it's injected and extracted: producer, consumer spans and links

Reviewer 4: > Why do we see CloudEvents as a messaging system?

We don't. That's why we asked to not define it in the messaging semantic conventions (which was the initial PR). As @lmolkova pointed out though, it's heavily used in the context of messaging systems.

> I would not even mention messaging. I'm still not sure if there is much overlap.

I'm fine with that too. It might be better to not have disguised TODO statements in the spec. Work that is left to be done can be captured in issues. In this context, one could just add a note to https://github.com/open-telemetry/opentelemetry-specification/issues/1979.

Author: I removed those statements and what's currently not covered.

**Status**: [Experimental](../../document-status.md)

<!-- Re-generate TOC with `markdown-toc --no-first-h1 -i` -->

<!-- toc -->

- [Definitions](#definitions)
- [Overview](#overview)
- [Conventions](#conventions)
  * [Spans](#spans)
    + [Creation](#creation)
    + [Processing](#processing)
  * [Attributes](#attributes)

<!-- tocstop -->

## Definitions

 From the

>  [CloudEvents specification:](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#overview)

Reviewer 5: ```suggestion
 [CloudEvents specification](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md#overview):
```

      | Producer | ---------------> | Intermediary |                                                                           
      +----------+                  +--------------+                                                                           
                                           |                                                                                   
                                           |                                                                                   
                                           |                                                                                   
                                           v                                                                                   
      +----------+                    +----------+                                                                             
      | Consumer | <----------------- |  Queue   |                                                                             
      +----------+                    +----------+ 
```

With the traditional parent-child trace model, the above scenario would produce

> two traces, completely independent from each other. It is not possible to
> correlate a producer with a consumer(s).

Reviewer 5: ```suggestion
two traces, completely independent from each other because the consumer
starts receiving (and thus has to specify a parent span) before it receives the event.
It is not possible to correlate a producer with a consumer(s) solely via a parent-child relationship.
```

|    |Link    +------------------------------------+    |
|    |                                                  |
|    |                                                  |
|    |                                                  |
|    |     Trace 2                                      |
|    |                                                  |
|    |     +---------------------------------------+    |
|    |     | Consumer: Receive (auto-instr)        |    |
|    |     +---------------------------------------+    |
|    |       +-------------------------------------+    |
|    +------ | Consumer: Process                   |    |
|            +-------------------------------------+    |
|                                                       |
+-------------------------------------------------------+
```

## Conventions

To achieve the trace above, it is necessary to capture the context of
the event creation so that when the CloudEvent reaches its destination(s), this
context can be continued. Each CloudEvent acts then as the medium of this
context propagation.

This document relies on the CloudEvents specification, which defines this
context propagation mechanism via the

> [CloudEvents Distributed Tracing Extension](https://github.com/cloudevents/spec/blob/v1.0.1/extensions/distributed-tracing.md).

Reviewer 5: ```suggestion
[CloudEvents Distributed Tracing Extension](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/extensions/distributed-tracing.md).
```
Since you reference 1.0.2 above as well (the files were reorganized in the meantime).

Once the trace context is set on the event
via the Distributed Tracing Extension, it MUST not be modified.

The remainder of this section describes the semantic conventions for Spans
required to achieve the proposed trace.

### Spans

#### Creation

Instrumentation SHOULD create a new span and populate the

> [CloudEvents Distributed Tracing Extension](https://github.com/cloudevents/spec/blob/v1.0.1/extensions/distributed-tracing.md)

Reviewer 5: ```suggestion
[CloudEvents Distributed Tracing Extension](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/extensions/distributed-tracing.md)
```

### Processing

When instrumented library supports processing of a single CloudEvent, instrumentation SHOULD create a new span to trace it. Instrumentation SHOULD set the remote trace context from the Distributed Tracing Extension as a link on the processing span.

**Span name:** `CloudEvents Process <event_type>`
**Span kind:** CONSUMER

> Note: CloudEvents processing does not follow a common pattern. Some "CloudEvents-enabled" libraries expose handlers for processing (e.g.  [CloudEvents Go SDK](https://github.com/cloudevents/sdk-go#receive-your-first-cloudevent)) and can auto-instrument the processing calls. Alternatively, a library might offer a handler, but the user chooses not to use it, or a library does not offer one at all. In such cases, the instrumentation responsibility falls under the application owner.

Reviewer 4: ```suggestion
Note: CloudEvents processing does not follow a common pattern. Some "CloudEvents-enabled" libraries expose handlers for processing (e.g.  [CloudEvents Go SDK](https://github.com/cloudevents/sdk-go#receive-your-first-cloudevent)) and can auto-instrument the processing calls. Alternatively, a library might offer a handler, but the user chooses not to use it, or a library does not offer one at all. In such cases, the instrumentation responsibility lies with the application owner.
```

I guess "falls on" would work too.

a span and MUST NOT modify the Distributed Tracing Extension on the event.

**Span name:** `CloudEvents Create <event_type>`

> **Span kind:** PRODUCER

Reviewer 5: This renders just like the h4 for Creation and Processing. You could make the hierarchy clearer by making it italic instead of bold or by indenting it via bullet points, for example.

Author: I will go with the bullet points approach, that renders better I think.

> Distributed Tracing Extension as a link on the processing span.

Reviewer 5: One might wonder which attributes to use. If the spec would say "with no attributes" it would prohibit adding them but you could say "with optional attributes" or "MAY add attributes" to make clear they are not specified.

Author: OK - I misunderstood and you were talking about attributes when adding the Link. I also addressed this on 714543a8963eea5f7cf7ebbe8efa29f52a3dae78
